containers/hawkular-metrics-openshift-integration/src/main/java/org/hawkular/openshift/auth/OpenShiftTokenAuthentication.java

```diff
@@ -0,0 +1,180 @@
+import org.codehaus.jackson.node.ObjectNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author mwringe
+ */
+public class OpenShiftTokenAuthentication {
+
+    private static final Logger logger = LoggerFactory.getLogger(OpenShiftTokenAuthentication.class);
+
+    private static final String HAWKULAR_TENANT = "hawkular-tenant";
+
+    private static final String KUBERNETES_MASTER_URL =
+            System.getProperty("KUBERNETES_MASTER_URL", "https://kubernetes.default.svc.cluster.local");
+
+    //The resource to check against for security purposes. For this version we are allowing Metrics based on a users
+    //access to the pods in in a particular project.
+    private static final String RESOURCE = "pods";
+
+    private static final String KIND = "SubjectAccessReview";
+
+    enum HTTP_METHOD {GET, PUT, POST, DELETE, PATCH};
+
+    public void doFilter(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws
+            IOException, ServletException {
+
+        String token = request.getHeader(OpenShiftAuthenticationFilter.AUTHORIZATION_HEADER);
+        String tenant = request.getHeader(HAWKULAR_TENANT);
+
+        if (token == null || tenant == null) {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST,
+                               "The '" + OpenShiftAuthenticationFilter.AUTHORIZATION_HEADER + "' and '"
+                                       + HAWKULAR_TENANT + "' headers" + " are required");
+            return;
+        }
+
+        if (isAuthorized(request.getMethod(), token, tenant)) {
+            filterChain.doFilter(request, response);
+            return;
+        } else {
+            response.sendError(HttpServletResponse.SC_FORBIDDEN);
+            return;
+        }
+    }
+
+    private boolean isAuthorized(String method, String token, String projectId) {
+        try {
+
+            String verb = getVerb(method);
+
+            String pod_namespace = System.getenv("POD_NAMESPACE");
+            if (pod_namespace == null) {
+                pod_namespace = "default";
+            }
+
+            String path = "/oapi/v1/namespaces/" + pod_namespace + "/subjectaccessreviews";
+            URL url = new URL(KUBERNETES_MASTER_URL + path);
+
+            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
+
+            //Configure the outgoing request
+            connection.setRequestMethod("POST");
+            connection.setDoOutput(true);
+            //Set Headers
+            connection.setRequestProperty("Accept", "application/json");
+            connection.setRequestProperty("Authorization", token);
+
+            //Add the body
+            OutputStream outputStream = connection.getOutputStream();
+            for (byte b : generateSubjectAccessReview(projectId, verb).getBytes()) {
+                outputStream.write(b);
+            }
+            outputStream.flush();
+            outputStream.close();
```

Comment:
If IOException is thrown by the flush() or write(), this close() is never called. Refactor to use try-with or at least in the finally block.